docs/container-vm-image-wrapper.md

```diff
@@ -0,0 +1,238 @@
+mechanism used by pods. To do this, KubeVirt must have a way of storing and
+accessing VM images using a container registry.
+
+## Requirements
+* Users must have a workflow for pushing VM images into a container registry.
+* KubeVirt runtime must be able to consume VM images stored in the container
+  registry.
+
+## Initial Use Cases
+For the sake of simplicity, this proposal is focusing on the details involved
+with use cases where VMs are launched using ephemeral storage that is not
+persistent between VM restarts.
+
+It is possible to extend usage of container registry backed VM images to use
+cases where VMs are stateful and need live migration support. This functionality
+can be built upon the foundation laid by this initial proposal. High level notes
+on potential directions for accomplishing this are provided at the end of this
+document.
+
+### VMs with Ephemeral Local Storage
+
+Users want to launch VM workloads backed by local ephemeral storage. The VM
+workload does not need to remain persistent across VM restarts and the workload
+does not require live migration support.
+
+### VM Replication with Ephemeral Local Storage.
+
+Users want to horizontally scale VM workloads backed by local ephemeral
+storage. *Pending introduction and implementation of VMGroup object*
+
+## Implementation 
+
+### Concepts
+
+When a container starts, the container's image is pulled from a remote
+container registry. Each container image consists of multiple layers and only
+the layers that are not already present locally are retrieved from the remote
+registry. 
+
+When a container is being initialized, a union mount filesystem is used to
+combine all the container image's layers together to form the container's
+directory tree. When processes in a container modify contents on the filesystem,
+the processes are not modifying the layers provided by the container image, they
+are modifying a layer of the union filesystem that is tracking the delta between
+the active container and the underlying layers provided by the container image.
+
+The delta between the container image and an active container is ephemeral and
+is automatically freed when container is destroyed. Depending on the union
+mount filesystem implementation used, the delta may be tracked at the block
+level or the file level. When tracked at the file level, the first time a file
+is written to causes the entire file to be copied into the delta layer.
+
+These are the concepts we're building upon to provide an ephemeral disk to a VM
+backed backed by a container image.
+
+### High Level Design
+
+**Standardised KubeVirt VM Image Wrapper**
+
+KubeVirt provides a standardized base wrapper container image that serves up a
+user provided VM image as a network disk consumable by Libvirt. This base
+container image does not have a VM image inside of it, it merely contains
+everything necessary for serving up a VM image in a way Libvirt can consume.
```

Comment:
This is an interesting idea, but I think it has a serious flaw when considering software updates. If you need to update the kubevirt software, this will include updating the daemon that lives in this image. When we do this, every single VM image needs to be rebuilt to point to the new base image. IMHO this makes the approach impractical. We need to be able to update the software without invalidating content the user has provided